File: app/src/debug/java/org/tasks/DebugNetworkInterceptor.java
package org.tasks;

import android.content.Context;
import com.facebook.flipper.android.AndroidFlipperClient;
import com.facebook.flipper.plugins.network.FlipperOkhttpInterceptor;
import com.facebook.flipper.plugins.network.NetworkFlipperPlugin;
import com.google.api.client.http.HttpRequest;
import com.google.api.client.http.HttpResponse;
import java.io.IOException;
import javax.inject.Inject;
import okhttp3.OkHttpClient;
import org.tasks.injection.ForApplication;

public class DebugNetworkInterceptor {

  private final Context context;

  @Inject
  public DebugNetworkInterceptor(@ForApplication Context context) {
    this.context = context;
  }

  private static NetworkFlipperPlugin getNetworkPlugin(Context context) {
    return AndroidFlipperClient.getInstance(context).getPlugin(NetworkFlipperPlugin.ID);
  }

  public void add(OkHttpClient.Builder builder) {
    builder.addNetworkInterceptor(new FlipperOkhttpInterceptor(getNetworkPlugin(context)));
  }

  public <T> T execute(HttpRequest request, Class<T> responseClass) throws IOException {
    FlipperHttpInterceptor<T> interceptor =
        new FlipperHttpInterceptor<>(getNetworkPlugin(context), responseClass);
    request
        .setInterceptor(new ChainedHttpExecuteInterceptor(request.getInterceptor(), interceptor))
        .setResponseInterceptor(interceptor)
        .execute();
    return interceptor.getResponse();
  }

  public <T> T report(HttpResponse httpResponse, Class<T> responseClass, long start, long finish)
      throws IOException {
    FlipperHttpInterceptor<T> interceptor =
        new FlipperHttpInterceptor<>(getNetworkPlugin(context), responseClass);
    interceptor.report(httpResponse, start, finish);
    return interceptor.getResponse();
  }
}

File: app/src/debug/java/org/tasks/DebugNetworkInterceptor.kt
package org.tasks

import android.content.Context
import com.facebook.flipper.android.AndroidFlipperClient
import com.facebook.flipper.plugins.network.FlipperOkhttpInterceptor
import com.facebook.flipper.plugins.network.NetworkFlipperPlugin
import com.google.api.client.http.HttpRequest
import com.google.api.client.http.HttpResponse
import dagger.hilt.android.qualifiers.ApplicationContext
import okhttp3.OkHttpClient
import java.io.IOException
import javax.inject.Inject

class DebugNetworkInterceptor @Inject constructor(@param:ApplicationContext private val context: Context) {
    fun apply(builder: OkHttpClient.Builder?) {
        builder?.addNetworkInterceptor(FlipperOkhttpInterceptor(getNetworkPlugin(context)))
    }

    @Throws(IOException::class)
    fun <T> execute(request: HttpRequest, responseClass: Class<T>): T? {
        val interceptor = FlipperHttpInterceptor(getNetworkPlugin(context), responseClass)
        request
                .setInterceptor(interceptor)
                .setResponseInterceptor(interceptor)
                .execute()
        return interceptor.response
    }

    @Throws(IOException::class)
    fun <T> report(httpResponse: HttpResponse, responseClass: Class<T>, start: Long, finish: Long): T? {
        val interceptor = FlipperHttpInterceptor(getNetworkPlugin(context), responseClass)
        interceptor.report(httpResponse, start, finish)
        return interceptor.response
    }

    private fun getNetworkPlugin(context: Context): NetworkFlipperPlugin {
        return AndroidFlipperClient.getInstance(context).getPlugin(NetworkFlipperPlugin.ID)!!
    }
}